crucible-llvm/src/Lang/Crucible/LLVM/SymIO.hs

```diff
@@ -0,0 +1,560 @@
+        -- handles mapping the underlying FileHandle to an int file descriptor
+        -- internally.
+        let toLit = W4.Char8Literal . TextEncoding.encodeUtf8 . SymIO.fdTargetToText
+        when (Map.member SymIO.StdinTarget (SymIO.concreteFiles initContents) || Map.member SymIO.StdinTarget (SymIO.symbolicFiles initContents)) $ do
+          stdinFilename <- liftIO $ W4.stringLit sym (toLit SymIO.StdinTarget)
+          _inFD <- SymIO.openFile' fsVar stdinFilename >>= allocateFileDescriptor llfs
+          return ()
+        when (SymIO.useStdout initContents) $ do
+          stdoutFilename <- liftIO $ W4.stringLit sym (toLit SymIO.StdoutTarget)
+          _outFD <- SymIO.openFile' fsVar stdoutFilename >>= allocateFileDescriptor llfs
+          return ()
+        when (SymIO.useStderr initContents) $ do
+          stderrFilename <- liftIO $ W4.stringLit sym (toLit SymIO.StderrTarget)
+          _errFD <- SymIO.openFile' fsVar stderrFilename >>= allocateFileDescriptor llfs
+          return ()
+
+  return (llfs, globals1, SomeOverrideSim bootstrapStdio)
+
+type FDescMapType w = IntrinsicType "LLVM_fdescmap" (EmptyCtx ::> BVType w)
+
+instance (IsSymInterface sym) => IntrinsicClass sym "LLVM_fdescmap" where
+  type Intrinsic sym "LLVM_fdescmap" (EmptyCtx ::> BVType w) = FDescMap sym w
+
+  muxIntrinsic sym _iTypes _nm (Empty :> (BVRepr _w)) = muxFDescMap sym
+  muxIntrinsic _ _ nm ctx = \_ _ _ -> typeError nm ctx
+
+pattern FDescMapRepr :: () => (1 <= w, ty ~ FDescMapType w) => PN.NatRepr w -> TypeRepr ty
+pattern FDescMapRepr w <- IntrinsicRepr (PC.testEquality (PS.knownSymbol @"LLVM_fdescmap") -> Just PC.Refl) (Empty :> BVRepr w)
+  where
+    FDescMapRepr w = IntrinsicRepr PS.knownSymbol (Empty :> BVRepr w)
+
+muxFDescMap
+  :: IsSymInterface sym
+  => sym
+  -> W4.Pred sym
+  -> FDescMap sym ptrW
+  -> FDescMap sym ptrW
+  -> IO (FDescMap sym ptrW)
+muxFDescMap sym p (FDescMap nextT mapT) (FDescMap nextF mapF) = do
+  let
+    keys = Set.toList $ Set.union (Map.keysSet mapT) (Map.keysSet mapF)
+    next = max nextT nextF
+  fmap (FDescMap next . Map.fromList) $ forM keys $ \k -> do
+    let vT = W4P.joinMaybePE (Map.lookup k mapT)
+    let vF = W4P.joinMaybePE (Map.lookup k mapF)
+    r <- mergePartExpr sym (CMT.mergeMuxTree sym) p vT vF
+    return (k,r)
+
+-- | The intrinsics supporting symbolic I/O in LLVM
+--
+-- Note that this includes the base intrinsic types from crucible-symio, so
+-- those do not need to be added again.
+llvmSymIOIntrinsicTypes :: IsSymInterface sym => IntrinsicTypes sym
+llvmSymIOIntrinsicTypes = id
+  . MapF.insert (PS.knownSymbol :: PS.SymbolRepr "LLVM_fdescmap") IntrinsicMuxFn
+  $ SymIO.symIOIntrinsicTypes
+
+-- | Resolve a symbolic file descriptor to a known allocated file handle.
+-- The partial result is undefined if the descriptor is not found in the
+-- file handle table.
+getHandle
+  :: forall sym ptrW
+   . IsSymInterface sym
+  => sym
+  -> W4.SymNat sym
```

Comment:
Would it make more sense to model FDs as a muxtree instead?

Reply:
In this particular case I'm wondering why it is a Nat instead of a Bitvector. Did you mean for this function on the internals, or in general? In general, I'm a bit worried about the `FD_SET` family of macros for e.g., `select` and `poll` that might be assuming bitvector operations are possible (it is a bit hard to tell).

Reply:
Not sure.  It looks like `fdset` are generally implemented as a bitset array, with size expected to be no more than `FD_SETSIZE`.

As far as I can tell, the only call side for `getHandle` has immediately previously coerced a bitvector into a natural.  Better not to do this if possible, that translation really sucks for symbolic values.

Reply:
Yup, let me see if I can change this one to just be a bitvector at the very least

Reply:
Okay, I switched it to be a SymBV instead of a SymNat